src/layout_engine.rs

```diff
@@ -0,0 +1,206 @@
+//! The layout engine converts the layout tree to rectangles of fixed sizes on the screen,
+//! in order to know how to draw the various components.
+
+use crate::layout_abstract_syntax::*;
+use crate::layout_abstract_types::*;
+
+/// An axis along which layout is performed
+#[derive(Copy, Clone, Debug, Eq, PartialEq)]
+enum Axis {
+	/// Horizontal axis, which is normally used by rows.
+	Horizontal,
+	/// Vertical axis, which is normally used by columns.
+	Vertical,
+}
+
+impl Axis {
+	/// Returns the axis perpendicular to this one.
+	fn perpendicular(&self) -> Axis {
+		match self {
+			Axis::Horizontal => Axis::Vertical,
+			Axis::Vertical => Axis::Horizontal,
+		}
+	}
+}
+/// Built-in component attributes used during layout.
+#[derive(Debug)]
+struct LayoutAttributes {
+	pub width: Dimension,
+	pub height: Dimension,
+}
+
+impl LayoutAttributes {
+	/// Retrieves the size of this element along the specified axis.
+	fn size_along(&self, axis: Axis) -> Dimension {
+		match axis {
+			Axis::Horizontal => self.width,
+			Axis::Vertical => self.height,
+		}
+	}
+
+	// fn alignment_along(axis: Axis)
+}
+
+/// A dimension is a measure along an axis.
+#[derive(Copy, Clone, Debug, PartialEq)]
+enum Dimension {
+	/// Absolute value in pixels.
+	AbsolutePx(f32),
+	/// Percent of parent container size along the same axis.
+	Percent(f32),
+	/// Percent of free space remaining in parent container.
+	PercentRemainder(f32),
+	/// Minimum size required to fit the children.
+	Inner,
+	/// Size relative to the width of this component.
+	Width,
+	/// Size relativ to the height of this component.
+	Height,
+}
+
+impl Dimension {
+	/// Converts an `AttributeValue` to a `Dimension`.
+	fn from_attribute_value(value: &AttributeValue) -> Self {
+		if let AttributeValue::TypeValue(values) = value {
+			assert_eq!(values.len(), 1, "Unexpected number of values");
```

Comment:
It might be more useful for this function to return a vector of `Dimension`s. It would make sense to treat everything as multi-value attributes (for example `padding="1px 2px 3px 4px"`), and single-value attributes are multi-value attributes of length 1.

Reply:
I'll leave that as a to-do for the future